### ISSUE TITLE
__fish_complete_suffix is busted

### ISSUE THREAD

**Issue by floam** [fish-shell/fish-shell#8512]
<img width="424" alt="Screen Shot 2021-12-01 at 6 56 10 AM" src="https://user-images.githubusercontent.com/291142/144258250-2de1af79-28af-46a3-91f6-e56532354aca.png">

**Comment by floam** [comment]
It seems this might have been intentional - the idea being to use it with --keep-order and only put the mentioned suffix at the top of the list. When not used with `-k` it does nothing at all, then, however. They right way to achieve that effect should have been to kept the prior behavior of this function, outputting, which outputs just the matching filenames, and using it in concert with --force-files or not with --no-files to get the rest, no?

**Comment by krobelus** [comment]
When reverting bae75c96d945504ff1e13464149175135e402a4f and allowing files:

```diff
diff --git a/share/completions/unzip.fish b/share/completions/unzip.fish
index fe513b87c..cfd58b857 100644
--- a/share/completions/unzip.fish
+++ b/share/completions/unzip.fish
@@ -41,7 +41,7 @@ if unzip -v 2>/dev/null | string match -eq Debian
 else
 
     # all tokens should be zip files
-    complete -c unzip -k -xa '(
+    complete -c unzip -k -ra '(
         __fish_complete_suffix .zip
         __fish_complete_suffix .jar
         __fish_complete_suffix .aar
```

then we get all files (zip or not) in lexical order, making `__fish_complete_suffix` redundant.

I guess we could use a tiered approach, something like

```diff
diff --git a/share/completions/unzip.fish b/share/completions/unzip.fish
index fe513b87c..5246068c3 100644
--- a/share/completions/unzip.fish
+++ b/share/completions/unzip.fish
@@ -41,10 +41,19 @@ if unzip -v 2>/dev/null | string match -eq Debian
 else
 
     # all tokens should be zip files
-    complete -c unzip -k -xa '(
+    complete -c unzip -n '
+        __fish_complete_suffix .zip ||
+        __fish_complete_suffix .jar ||
+        __fish_complete_suffix .aar
+    ' -xa '(
         __fish_complete_suffix .zip
         __fish_complete_suffix .jar
         __fish_complete_suffix .aar
     )'
+    complete -c unzip -n '
+        not __fish_complete_suffix .zip &&
+        not __fish_complete_suffix .jar &&
+        not __fish_complete_suffix .aar
+    ' --force-files
 
 end
diff --git a/share/functions/__fish_complete_suffix.fish b/share/functions/__fish_complete_suffix.fish
index 5d0402dd8..694f788e0 100644
--- a/share/functions/__fish_complete_suffix.fish
+++ b/share/functions/__fish_complete_suffix.fish
@@ -102,4 +102,5 @@ function __fish_complete_suffix -d "Complete using files"
         printf "%s$desc\n" $files
     end
 
+    set -q files[1]
 end
```

But I don't think we do that anywhere.

An alternative fix is to revert and have fish fall back to file completions if custom completions fail.
I believe this has been brought up a few times, maybe it's worth trying?

**Comment by faho** [comment]
>An alternative fix is to revert and have fish fall back to file completions if custom completions fail.

Not giving files when other things fail is a *biiiig* usability win, especially for things like e.g. `git thing<TAB>`.

This was a really really good change, and we definitely do not want to revert it.

An alternative is to add a secondary mode `--or-files` that gives files if that specific completion doesn't list anything, *but* __fish_complete_suffix typically isn't authoritative, most commands do not actually care about the suffix, and files can be named whatever the user wants. So the sorting approach is fundamentally sensible.

**Comment by floam** [comment]
Fair enough - I'm not really digging the other solutions proposed. Gunzip for example on my system though does care about suffixes, and the prior behavior there was truly better. Also seems like the sorting isn't working for `clang`, not sure what is up with that.

**Comment by EmilySeville7cfg** [comment]
![image](https://user-images.githubusercontent.com/42812113/144527380-441c645d-3ee6-4b95-984b-139144dbfaea.png)

> An alternative fix is to revert and have fish fall back to file completions if custom completions fail.

Maybe better to fall back only for `__fish_complete_suffix`?

**Comment by zanchey** [comment]
> An alternative is to add a secondary mode `--or-files` that gives files if that specific completion doesn't list anything

That seems useful to me.

**Comment by faho** [comment]
> What is the recommended way to handle commands that do require a specific suffix?

`complete --no-files` and generate the list of files yourself.

In our experience that basically does not happen. E.g. even with your use case of kernel modules you might see ".ko", but they could also be ".ko.gz" or ".ko.xz" or whatever other compression format someone decides to add. Requiring a suffix means you get to keep squashing bugs whenever there's a new suffix, so offering *too many* files is preferable over not offering some valid files.

**Comment by krobelus** [comment]
> What is the recommended way to handle commands that do require a specific suffix?

maybe `__fish_complete_suffix .log | grep -E '\.log($|\\\\t)'`

**Comment by krobelus** [comment]
> it seems to scale poorly with multiple extensions 

no, use `grep -E`

> doesn't list directories

those end with a /
